Overhaul pdfobj

cormacrelf opened the pull request:
Depends on #175

- :tada: PdfDict uses an insertion-order-preserving hash map, from `indexmap` :tada:
- PdfName uses CString under the hood
- Streams and arrays get a much more convenient API, and more things
  operate directly on streams
- Rustified filter_decoded

Note that review is a bit annoying as I abstracted away `*mut pdf_obj` behing a type alias, `PdfObjRef`. You can basically click 'viewed' on every file except `dpx_pdfobj.rs`.

cormacrelf committed:
dpx_pdfobj ergonomics

cormacrelf committed:
more

cormacrelf committed:
Mostly rustify filter_decoded

cormacrelf committed:
Use Box to allocate pdf_obj; more stream API

cormacrelf committed:
Hide *mut pdf_obj behing PdfObjRef alias

cormacrelf committed:
safe_new_obj for constructing pdf_obj without needing unsafe

cormacrelf committed:
safely initialize pdf_name

cormacrelf committed:
rename pdf_name

cormacrelf committed:
Convert PdfName to use CString properly

cormacrelf committed:
More Box allocation for adding to a pdf_dict

cormacrelf committed:
Use hashmap for pdf_dict (now non-deterministic)

cormacrelf committed:
use indexmap for pdf_dict so it's deterministic again

cormacrelf committed:
Rename pdf_dict -> PdfDict

cormacrelf committed:
Restore some lost documentation

chirsz-ever commented:
Do we have a plan to use `String` everywhere in the near future?

cormacrelf commented:
We need to provide `&'a CStr` sometimes, so that's a non-local change.

burrbull commented:
> Do we have a plan to use `String` everywhere in the near future?

Pdf is byte-oriented format, so:
```
*const i8 -> CString(CStr) -> Vec<u8>(&[u8])
```
`String(&str)` where it is appropriate

burrbull commented:
`lopdf` uses `LinkedHashMap`

cormacrelf commented:
> `lopdf` uses `LinkedHashMap`

`linked-hash-map` is unmaintained, but they do the same thing, so I went for `indexmap`. It only takes a few moments to swap them over. I'll push and see what happens to the tests.

cormacrelf commented:
Ok, it's much worse. Switching back.

mulimoen commented:
I thing the hashmap can hold multiple values with the same key. You could try `multimap`

cormacrelf commented:
@mulimoen Hm? The old linked list code would loop until it found the same key, and replace it, and pdf_obj_release the existing value. If it didn't find any, it would tack it on the end. There were never any dupes. IndexMap has the same properties. It's just a hashmap that preserves insertion order.

mulimoen commented:
@cormacrelf I assumed it was using `ht_lookup_table`, but seems I was wrong.

cormacrelf committed:
Make filter_decoded work properly for PNG decoding again

(Takes on `png` dependency -- this will be shared by `image`, eventually, hopefully.)

cormacrelf commented on dpx/src/dpx_pdfobj.rs:
```diff
@@ -1110,159 +1152,125 @@ impl Drop for pdf_array {
+    let mut make_key_ptr = |key: &PdfName| safe_new_obj(PdfObjType::NAME, |object| {
+        object.data = Box::into_raw(Box::new(key.clone())) as *mut libc::c_void;
+    });
+    dict.foreach_dict(|k, v, pdata| proc(make_key_ptr(k), v, pdata), pdata)
```
This is the only thing I can find that's really different --

* The original calls proc with the actual PdfObjRef (which happens to be a PdfName)
* This code makes a new PdfObjRef to call proc with, and also happens to never drop it.

cormacrelf committed:
free tmp_key_ptr in pdf_foreach_dict

cormacrelf committed:
fix scanline vecs being to short in filter_decoded

cormacrelf commented:
Ok, I think that's going to pass, we'll see in 125 minutes. But I'm mostly happy.

cormacrelf commented:
Well, if anyone can figure out what's wrong, let me know, otherwise I'll try bisecting into multiple PRs.

burrbull commented:
It's better to split this PR on several in any case.

burrbull commented on dpx/src/dpx_pdfobj.rs:
```diff
@@ -123,9 +129,49 @@ impl pdf_obj {
     pub fn is_array(&self) -> bool {
         PdfObjType::from(self.typ) == PdfObjType::ARRAY
     }
+    pub fn get_array_mut(&mut self) -> &mut PdfArray {
```
maybe better `as_array_mut` like in `lopdf`?

cormacrelf replied:
Agree
